### PR TITLE
[rdf-js] Add quad generic to DataFactory

### PR DESCRIPTION
**Pull request by thewilkybarkid**
Adds the quad type to `DataFactory` as a generic, rather than having it on its functions. This allows matching a `DataFactory` up with how it's used, such as https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2eccf3b0aa829ea8570b4682aba395a2bf154bc7/types/rdf-dataset-indexed/index.d.ts#L8

I've tried to update its usage as well.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: n/a
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by thewilkybarkid** [comment]
In draft as I haven't added tests, but will wait on a green light before going through them.

File: types/rdf-ext/lib/Dataset.d.ts [hunk 1]
```diff
@@ -1,13 +1,13 @@
-import { Dataset } from 'rdf-js';
+import { BaseQuad, Dataset } from 'rdf-js';
 import { PropType } from './_PropType';
 import QuadExt = require('./Quad');
 
-interface DatasetExt extends Dataset {
+interface DatasetExt<Q extends BaseQuad = QuadExt> extends Dataset<Q> {
   readonly length: number;
   toJSON(): Array<ReturnType<PropType<QuadExt, 'toJSON'>>>;
```
Comment (thewilkybarkid) [inline types/rdf-ext/lib/Dataset.d.ts]: Because of this line, does this mean it should be `Q extends QuadExt`?

Comment (tpluscode) [reply types/rdf-ext/lib/Dataset.d.ts]: I think it should be a closed type

```
interface DatasetExt extends Dataset<QuadExt> 
```

rdf-ext will always use its `Ext` types AFAICT

Comment (thewilkybarkid) [reply types/rdf-ext/lib/Dataset.d.ts]: Same for https://github.com/DefinitelyTyped/DefinitelyTyped/blob/1163ead296d84e7a3c80d71e7c81ecbd1a130e9a/types/rdf-ext/lib/DataFactory.d.ts#L42-L43 too?

Comment (tpluscode) [reply types/rdf-ext/lib/Dataset.d.ts]: I should think that it does not make sense to make anything in rdf-ext generic. Those implementations as specific specialisations of the base interface

**Comment by thewilkybarkid** [comment]
Holiday rather than abandoned. 😅

But happy to carry on in #41651.

**Pull request closed without merging**